autorest/azure/auth/auth.go

```diff
@@ -41,59 +41,105 @@ import (
 // 3. Username password
 // 4. MSI
 func NewAuthorizerFromEnvironment() (autorest.Authorizer, error) {
-	tenantID := os.Getenv("AZURE_TENANT_ID")
-	clientID := os.Getenv("AZURE_CLIENT_ID")
-	clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
-	certificatePath := os.Getenv("AZURE_CERTIFICATE_PATH")
-	certificatePassword := os.Getenv("AZURE_CERTIFICATE_PASSWORD")
-	username := os.Getenv("AZURE_USERNAME")
-	password := os.Getenv("AZURE_PASSWORD")
-	envName := os.Getenv("AZURE_ENVIRONMENT")
-	resource := os.Getenv("AZURE_AD_RESOURCE")
-
-	var env azure.Environment
+	settings := getEnvVars()
+
+	env, err := getEnv(settings.envName)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.resource == "" {
+		settings.resource = env.ResourceManagerEndpoint
+	}
+
+	return getAuth(settings, env)
+}
+
+// NewKeyvaultAuthorizerFromEnvironment creates a keyvault dataplane Authorizer configured from environment variables in the order:
+// 1. Client credentials
+// 2. Client certificate
+// 3. Username password
+// 4. MSI
+func NewKeyvaultAuthorizerFromEnvironment() (autorest.Authorizer, error) {
+	settings := getEnvVars()
+	settings.resource = os.Getenv("AZURE_KEYVAULT_RESOURCE")
+
+	env, err := getEnv(settings.envName)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.resource == "" {
+		settings.resource = strings.TrimSuffix(env.KeyVaultEndpoint, "/")
+	}
+
+	return getAuth(settings, env)
+}
+
+type envSettings struct {
+	tenantID            string
+	clientID            string
+	clientSecret        string
+	certificatePath     string
+	certificatePassword string
+	username            string
+	password            string
+	envName             string
+	resource            string
+}
+
+func getEnvVars() envSettings {
+	return envSettings{
+		tenantID:            os.Getenv("AZURE_TENANT_ID"),
+		clientID:            os.Getenv("AZURE_CLIENT_ID"),
+		clientSecret:        os.Getenv("AZURE_CLIENT_SECRET"),
+		certificatePath:     os.Getenv("AZURE_CERTIFICATE_PATH"),
+		certificatePassword: os.Getenv("AZURE_CERTIFICATE_PASSWORD"),
+		username:            os.Getenv("AZURE_USERNAME"),
+		password:            os.Getenv("AZURE_PASSWORD"),
+		envName:             os.Getenv("AZURE_ENVIRONMENT"),
+		resource:            os.Getenv("AZURE_AD_RESOURCE"),
+	}
+}
+
+func getEnv(envName string) (env azure.Environment, err error) {
 	if envName == "" {
 		env = azure.PublicCloud
 	} else {
-		var err error
 		env, err = azure.EnvironmentFromName(envName)
-		if err != nil {
-			return nil, err
-		}
-	}
-
-	if resource == "" {
-		resource = env.ResourceManagerEndpoint
 	}
+	return
+}
 
+func getAuth(settings envSettings, env azure.Environment) (autorest.Authorizer, error) {
```

Comment:
I think this should be a method on envSettings.

Reply:
Fixed